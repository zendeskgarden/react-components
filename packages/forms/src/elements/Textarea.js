/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import PropTypes from 'prop-types';
import useFieldContext from '../utils/useFieldContext';
import { StyledTextarea } from '../styled';
import VALIDATION from '../utils/validation';

/**
 * Must be rendered within a `<Field>` element; accepts all
 * `<textarea />` attributes and events.
 */
const Textarea = React.forwardRef((props, ref) => {
  const { getInputProps } = useFieldContext();

  return (
    <StyledTextarea
      {...getInputProps(
        {
          ref,
          ...props
        },
        { isDescribed: true }
      )}
    />
  );
});

Textarea.propTypes = {
  /** Apply compact styling */
  isCompact: PropTypes.bool,
  /** Remove borders and padding */
  isBare: PropTypes.bool,
  /** Apply inset `box-shadow` styling on focus */
  focusInset: PropTypes.bool,
  /** Display a mechanism for vertical resize */
  isResizable: PropTypes.bool,
  validation: PropTypes.oneOf([VALIDATION.SUCCESS, VALIDATION.WARNING, VALIDATION.ERROR])
};

export default Textarea;
