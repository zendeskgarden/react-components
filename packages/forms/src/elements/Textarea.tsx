/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { TextareaHTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import useFieldContext from '../utils/useFieldContext';
import { StyledTextarea } from '../styled';
import { VALIDATION } from '../utils/validation';

export interface ITextareaProps extends TextareaHTMLAttributes<HTMLTextAreaElement> {
  /** Apply compact styling */
  isCompact?: boolean;
  /** Remove borders and padding */
  isBare?: boolean;
  /** Apply inset `box-shadow` styling on focus */
  focusInset?: boolean;
  /** Display a mechanism for vertical resize */
  isResizable?: boolean;
  validation?: VALIDATION;
}

/**
 * Must be rendered within a `<Field>` element; accepts all
 * `<textarea />` attributes and events.
 */
export const Textarea = React.forwardRef<HTMLTextAreaElement, ITextareaProps>((props, ref) => {
  const fieldContext = useFieldContext();

  let combinedProps = {
    ref,
    ...props
  };

  if (fieldContext) {
    combinedProps = fieldContext.getInputProps(combinedProps, { isDescribed: true });
  }

  return <StyledTextarea {...(combinedProps as any)} />;
});

Textarea.propTypes = {
  isCompact: PropTypes.bool,
  isBare: PropTypes.bool,
  focusInset: PropTypes.bool,
  isResizable: PropTypes.bool,
  validation: PropTypes.oneOf(['success', 'warning', 'error'])
};
