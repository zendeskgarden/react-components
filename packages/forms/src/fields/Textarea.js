/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import PropTypes from 'prop-types';
import useFieldContext from '../utils/useFieldContext';
import { StyledTextarea } from '../styled';

const VALIDATION = {
  SUCCESS: 'success',
  WARNING: 'warning',
  ERROR: 'error'
};

/**
 * Accepts all `<textarea />` props
 */
const Textarea = React.forwardRef((props, ref) => {
  const { getInputProps } = useFieldContext();

  return (
    <StyledTextarea
      {...getInputProps(
        {
          'data-garden-id': 'forms.textarea',
          'data-garden-version': PACKAGE_VERSION,
          ref,
          ...props
        },
        { isDescribed: true }
      )}
    />
  );
});

Textarea.propTypes = {
  small: PropTypes.bool,
  bare: PropTypes.bool,
  disabled: PropTypes.bool,
  focused: PropTypes.bool,
  /** Applies inset `box-shadow` styling on focus */
  focusInset: PropTypes.bool,
  hovered: PropTypes.bool,
  resizable: PropTypes.bool,
  validation: PropTypes.oneOf([VALIDATION.SUCCESS, VALIDATION.WARNING, VALIDATION.ERROR])
};

export default Textarea;
