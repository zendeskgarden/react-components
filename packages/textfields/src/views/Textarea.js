/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import PropTypes from 'prop-types';
import styled from 'styled-components';
import classNames from 'classnames';
import { retrieveTheme, isRtl } from '@zendeskgarden/react-theming';
import TextStyles from '@zendeskgarden/css-forms/dist/text.css';

import Input from './Input';

const COMPONENT_ID = 'textfields.textarea';
const VALIDATION = {
  SUCCESS: 'success',
  WARNING: 'warning',
  ERROR: 'error'
};

/**
 * Accepts all `<textarea>` props
 */
const Textarea = styled(Input.withComponent('textarea')).attrs({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  className: props =>
    classNames(TextStyles['c-txt__input--area'], {
      [TextStyles['is-resizable']]: props.resizable,

      // RTL
      [TextStyles['is-rtl']]: isRtl(props)
    })
})`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

Textarea.propTypes = {
  small: PropTypes.bool,
  bare: PropTypes.bool,
  disabled: PropTypes.bool,
  focused: PropTypes.bool,
  hovered: PropTypes.bool,
  resizable: PropTypes.bool,
  validation: PropTypes.oneOf([VALIDATION.SUCCESS, VALIDATION.WARNING, VALIDATION.ERROR])
};

/** @component */
export default Textarea;
