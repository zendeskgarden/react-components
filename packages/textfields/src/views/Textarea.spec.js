/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { mount } from 'enzyme';
import { mountWithTheme } from '@zendeskgarden/react-testing';

import Textarea from './Textarea';

describe('Textarea', () => {
  it('renders default styling', () => {
    const wrapper = mount(<Textarea />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders RTL styling', () => {
    const wrapper = mountWithTheme(<Textarea />, { rtl: true });

    expect(wrapper).toMatchSnapshot();
  });

  it('renders small styling if provided', () => {
    const wrapper = mount(<Textarea small />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders bare styling if provided', () => {
    const wrapper = mount(<Textarea bare />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders disabled styling if provided', () => {
    const wrapper = mount(<Textarea disabled />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders focused styling if provided', () => {
    const wrapper = mount(<Textarea focused />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders hovered styling if provided', () => {
    const wrapper = mount(<Textarea hovered />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders resizable styling if provided', () => {
    const wrapper = mount(<Textarea resizable />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('validation', () => {
    ['success', 'warning', 'error', 'none'].forEach(validation => {
      it(`renders ${validation} styling if provided`, () => {
        const wrapper = mount(<Textarea validation={validation} />);

        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
