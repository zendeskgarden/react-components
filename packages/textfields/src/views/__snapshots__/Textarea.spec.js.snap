// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea renders RTL styling 1`] = `
<Textarea>
  <Input
    className="c-txt__input--area is-rtl "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
  >
    <textarea
      className="c-txt__input--area is-rtl c-txt__input is-rtl "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
    />
  </Input>
</Textarea>
`;

exports[`Textarea renders bare styling if provided 1`] = `
<Textarea
  bare={true}
>
  <Input
    bare={true}
    className="c-txt__input--area "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
  >
    <textarea
      className="c-txt__input--area c-txt__input c-txt__input--bare "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
    />
  </Input>
</Textarea>
`;

exports[`Textarea renders default styling 1`] = `
<Textarea>
  <Input
    className="c-txt__input--area "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
  >
    <textarea
      className="c-txt__input--area c-txt__input "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
    />
  </Input>
</Textarea>
`;

exports[`Textarea renders disabled styling if provided 1`] = `
<Textarea
  disabled={true}
>
  <Input
    className="c-txt__input--area "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
    disabled={true}
  >
    <textarea
      className="c-txt__input--area c-txt__input is-disabled "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
      disabled={true}
    />
  </Input>
</Textarea>
`;

exports[`Textarea renders focused styling if provided 1`] = `
<Textarea
  focused={true}
>
  <Input
    className="c-txt__input--area "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
    focused={true}
  >
    <textarea
      className="c-txt__input--area c-txt__input is-focused "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
    />
  </Input>
</Textarea>
`;

exports[`Textarea renders hovered styling if provided 1`] = `
<Textarea
  hovered={true}
>
  <Input
    className="c-txt__input--area "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
    hovered={true}
  >
    <textarea
      className="c-txt__input--area c-txt__input is-hovered "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
    />
  </Input>
</Textarea>
`;

exports[`Textarea renders resizable styling if provided 1`] = `
<Textarea
  resizable={true}
>
  <Input
    className="c-txt__input--area is-resizable "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
    resizable={true}
  >
    <textarea
      className="c-txt__input--area is-resizable c-txt__input "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
    />
  </Input>
</Textarea>
`;

exports[`Textarea renders small styling if provided 1`] = `
<Textarea
  small={true}
>
  <Input
    className="c-txt__input--area "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
    small={true}
  >
    <textarea
      className="c-txt__input--area c-txt__input c-txt__input--sm "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
    />
  </Input>
</Textarea>
`;

exports[`Textarea validation renders error styling if provided 1`] = `
<Textarea
  validation="error"
>
  <Input
    className="c-txt__input--area "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
    validation="error"
  >
    <textarea
      className="c-txt__input--area c-txt__input c-txt__input--error "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
    />
  </Input>
</Textarea>
`;

exports[`Textarea validation renders success styling if provided 1`] = `
<Textarea
  validation="success"
>
  <Input
    className="c-txt__input--area "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
    validation="success"
  >
    <textarea
      className="c-txt__input--area c-txt__input c-txt__input--success "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
    />
  </Input>
</Textarea>
`;

exports[`Textarea validation renders warning styling if provided 1`] = `
<Textarea
  validation="warning"
>
  <Input
    className="c-txt__input--area "
    data-garden-id="textfields.textarea"
    data-garden-version="version"
    validation="warning"
  >
    <textarea
      className="c-txt__input--area c-txt__input c-txt__input--warning "
      data-garden-id="textfields.textarea"
      data-garden-version="version"
    />
  </Input>
</Textarea>
`;
